contracts/examples/DXswapSwapRelayer.sol

```diff
@@ -0,0 +1,296 @@
+
+        address pair = _pair(tokenA, tokenB, factory);
+
+        require(pair != address(0), 'DXswapRelayer: INVALID_PAIR_ADDRESS');
+
+        orderIndex = _OrderIndex();
+        orders[orderIndex] = Order({
+            tokenIn: amountInTokenA > 0 ? tokenA : tokenB,
+            tokenOut: amountInTokenA > 0 ? tokenB : tokenA,
+            amountIn: amountInTokenA > 0 ? amountInTokenA : amountInTokenB,
+            priceTolerance: priceTolerance,
+            minReserveA: minReserveA,
+            minReserveB: minReserveB,
+            oraclePair: pair,
+            deadline: deadline,
+            maxWindowTime: maxWindowTime,
+            oracleId: 0,
+            factory: factory,
+            executed: false
+        });
+
+        (uint256 reserveA, uint256 reserveB, ) = IDXswapPair(pair).getReserves();
+
+        /* Create an oracle to calculate average price before swap */
+        uint256 windowTime = _consultOracleParameters(
+            amountInTokenA,
+            amountInTokenB,
+            reserveA,
+            reserveB,
+            maxWindowTime
+        );
+        orders[orderIndex].oracleId = oracleCreator.createOracle(windowTime, pair);
+        emit NewOrder(orderIndex);
+    }
+
+    function executeOrder(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(orderIndex < orderCount, 'DXswapRelayer: INVALID_ORDER');
+        require(!order.executed, 'DXswapRelayer: ORDER_EXECUTED');
+        require(oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_RUNNING');
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+
+        address tokenIn = order.tokenIn;
+        uint256 amountOut;
+        amountOut = oracleCreator.consult(
+            order.oracleId,
+            tokenIn == address(0) ? IDXswapRouter(dxSwapRouter).WETH() : tokenIn,
+            order.amountIn
+        );
+
+        uint256 minAmountOut = amountOut.sub(amountOut.mul(order.priceTolerance) / PARTS_PER_MILLION);
+
+        order.executed = true;
+
+        _swap(tokenIn, order.tokenOut, order.amountIn, minAmountOut, order.factory);
+
+        emit ExecutedOrder(orderIndex);
+    }
+
+    // Updates a price oracle and sends a bounty to msg.sender
+    function updateOracle(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+        require(!oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_ENDED');
+        uint256 amountBounty = GAS_ORACLE_UPDATE.mul(tx.gasprice).add(BOUNTY);
+
+        (uint256 reserveA, uint256 reserveB, ) = IDXswapPair(order.oraclePair).getReserves();
+        require(reserveA >= order.minReserveA && reserveB >= order.minReserveB, 'DXswapRelayer: RESERVE_TOO_LOW');
+
+        oracleCreator.update(order.oracleId);
```

Comment:
I'm wondering what happens here if the reserves have changed and in turn oracle parameters are potentially not safe anymore? I.e. someone removes liquidity without it being brought below minimum levels, but changing the oracle's window time as calculated in `_consultOracleParameters`. I would maybe think about consulting the oracle parameters once again and eventually update them for the next read (and potentially reset the update counter)?

Reply:
I think it would be ok indeed to consult the oracle and check whether window time has changed or not. Let's keep in minds that the same solution is used for DXswapRelayer contract, so if we consider the oracle might be not safe we will have to change also that one, @jpkcambridge?